panther_hardware_interfaces/src/panther_imu.cpp

```diff
@@ -0,0 +1,542 @@
+}
+
+bool PantherImuSensor::IsMagnitudeSynchronizedWithAccelerationAndGyration(
+  const geometry_msgs::msg::Vector3 & mag_compensated)
+{
+  return IsVectorFinite(mag_compensated);
+}
+
+void PantherImuSensor::SpatialDataCallback(
+  const double acceleration[3], const double angular_rate[3], const double magnetic_field[3],
+  const double timestamp)
+{
+  const auto timestamp_s = timestamp * 1e-3;
+
+  const auto mag_compensated = ParseMagnitude(magnetic_field);
+  const auto ang_vel = ParseGyration(angular_rate);
+  const auto lin_acc = ParseAcceleration(acceleration);
+
+  // Skip the data callback when IMU if not calibrated
+  if (!IsIMUCalibrated(mag_compensated)) {
+    return;
+  }
+
+  const float dt = timestamp_s - last_spatial_data_callback_time_s_;
+  last_spatial_data_callback_time_s_ = timestamp_s;
+
+  // Wait for the a magnitude and an acceleration to initialize the algorithm
+  if (
+    !algorithm_initialized_ &&
+    !IsMagnitudeSynchronizedWithAccelerationAndGyration(mag_compensated)) {
+    return;
+  }
+
+  if (!algorithm_initialized_ || params_.stateless) {
+    try {
+      ComputeInitialOrientation(mag_compensated, lin_acc, timestamp_s);
+    } catch (const std::runtime_error & e) {
+      RCLCPP_ERROR_STREAM(logger_, "Exception during algorithm initialization: " << e.what());
+    }
+    return;
+  }
+
+  if (!params_.stateless) {
+    if (IsMagnitudeSynchronizedWithAccelerationAndGyration(mag_compensated)) {
+      UpdateMadgwickAlgorithm(ang_vel, lin_acc, mag_compensated, dt);
+    } else {
+      UpdateMadgwickAlgorithmIMU(ang_vel, lin_acc, dt);
+    }
+  }
+
+  UpdateAllStatesValues(ang_vel, lin_acc);
+}
+
+void PantherImuSensor::SpatialAttachCallback()
+{
+  RCLCPP_INFO(logger_, "IMU has attached!");
+  imu_connected_ = true;
+  on_activate(rclcpp_lifecycle::State{});
+}
+
+void PantherImuSensor::SpatialDetachCallback()
+{
+  RCLCPP_WARN(logger_, "IMU has detached!");
+  imu_connected_ = false;
+  imu_calibrated_ = false;
+  algorithm_initialized_ = false;
+  SetStateValuesToNans();
+  on_deactivate(rclcpp_lifecycle::State{});
+}
+
+void PantherImuSensor::UpdateMadgwickAlgorithm(
+  const geometry_msgs::msg::Vector3 & ang_vel, const geometry_msgs::msg::Vector3 & lin_acc,
+  const geometry_msgs::msg::Vector3 & mag_compensated, const double dt)
+{
+  if (params_.use_mag) {
+    filter_->madgwickAHRSupdate(
+      ang_vel.x, ang_vel.y, ang_vel.z, lin_acc.x, lin_acc.y, lin_acc.z, mag_compensated.x,
+      mag_compensated.y, mag_compensated.z, dt);
+  } else {
+    filter_->madgwickAHRSupdateIMU(
+      ang_vel.x, ang_vel.y, ang_vel.z, lin_acc.x, lin_acc.y, lin_acc.z, dt);
+  }
+}
+
+void PantherImuSensor::UpdateMadgwickAlgorithmIMU(
+  const geometry_msgs::msg::Vector3 & ang_vel, const geometry_msgs::msg::Vector3 & lin_acc,
+  const double dt)
+{
+  filter_->madgwickAHRSupdateIMU(
+    ang_vel.x, ang_vel.y, ang_vel.z, lin_acc.x, lin_acc.y, lin_acc.z, dt);
+}
```

Comment:
I don't understand the division into two methods. Are we certain that their use is appropriate?

Reply:
There should not be `if (params_.use_mag) {` .